cl/citations/tests.py

```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf,</span><span> supra</span><span>, at '
+             '</span><span class="page">2</span></span><pre class="inline">'
+             '. foobar</pre>'),
+
+            # Second kind of supra citation (with volume)
+            ('asdf, 123 supra, at 2. foo bar',
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf,</span> <span class="volume">123</span>'
+             '<span> supra</span><span>, at </span><span class="page">2</span>'
+             '</span><pre class="inline">. foo bar</pre>'),
+
+            # Third kind of supra citation (sans page)
+            ('asdf, supra, foo bar',
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf,</span><span> supra</span></span><pre '
+             'class="inline">, foo bar</pre>'),
+
+            # Fourth kind of supra citation (with period)
+            ('asdf, supra. foo bar',
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf,</span><span> supra</span></span><pre '
+             'class="inline">. foo bar</pre>'),
+
+            # Supra citation across line break
+            ('asdf, supra, at\n99 (quoting foo)',
+             '<pre class="inline"></pre><span class="citation no-link"><span'
+             ' class="antecedent">asdf,</span><span> supra</span><span>, at\n'
+             '</span><span class="page">99</span> </span><pre class="inline">'
+             '(quoting foo)</pre>'),
+
+            # First kind of id. citation ("Id., at 123")
+            ('asdf, id., at 123. Lorem ipsum dolor sit amet',
+             '<pre class="inline">asdf, </pre><span class="citation no-link">'
+             'id.,<span class="after_tokens"> <span class="after_token">at'
+             '</span> <span class="after_token">123.</span> <span class="'
+             'after_token">Lorem</span> </span></span><pre class="inline">'
+             'ipsum dolor sit amet</pre>'),
+
+            # Second kind of id. citation ("Ibid.")
+            ('asdf, Ibid. Lorem ipsum dolor sit amet',
+             '<pre class="inline">asdf, </pre><span class="citation no-link">'
+             'Ibid.<span class="after_tokens"> <span class="after_token">Lorem'
+             '</span> <span class="after_token">ipsum</span> <span class="'
+             'after_token">dolor</span> </span></span><pre class="inline">'
+             'sit amet</pre>'),
+
+            # Id. citation across line break
+            ('asdf." Id., at 315.\n       Lorem ipsum dolor sit amet',
+             '<pre class="inline">asdf." </pre><span class="citation no-link">'
+             'Id.,<span class="after_tokens"> <span class="after_token">at'
+             '</span> <span class="after_token">315.</span>\n       <span'
```

Comment:
I'm confused why `315` shows up as an `after_token` here, but in the `supra`'s these show up with the `page` class?

Reply:
This is because for `id` and `ibid`'s I don't attempt to do any parsing of the adjacent tokens at all. Instead, I just grab the three `after_token`'s (whatever they are), and use them as an "anchor" for building a regex string that correctly matches the proper `id` or `ibid` reference in the HTML. And since `id` and `ibid` don't always have page numbers, I saw no reason to try parsing them specifically (since they won't help in the matching process, even if they were extracted).

For `supra`s, by contrast, those regex strings *are* built with page numbers as a specific component. So they are extracted for that.